### ISSUE TITLE
Feature idea: Language agnostic source code in non-plain text format

### ISSUE THREAD

**Issue by LeonG-ZA** [dotnet/roslyn#28232]
This isn't really important, but I've had this cool idea for years.

As we all know coding styles can differ a bit between teams and this can cause religious wars...
For example:

Some people like to put curly braces on the same line:
`public void SomeMethod() {`
Instead of:
`public void SomeMethod()`
`{`
There are some other examples, but you get the idea. Somebody might 'fix' the problem and check the code into source control.

What if code is not actually saved as plain text, but in some custom xml/json/binary format..
Every time somebody opens the file the IDE de-serializes the file and displays it correctly based on your coding style. You can check-in changes and it doesn't affect the other members who don't like your coding style. 

This can also be awesome to support non-English speakers around the world:
Every name, such as variables, class names, namespace names etc can have (optional) different names based on the language region selected in the IDE. This can be a simple drop down in the IDE to change the language on the fly.
If none are set for those names then the default English names are shown even when the language isn't English. The developer can the change those English names.
The same applies to comments. If none are set then the default English comments are shown and then changed.

One issue is that if exceptions are thrown then the Error/Exception at line number '...' would need to change to Error/Exception at 'section?' number '...'
Another issue is the references to different names located in other files. The default English would need to be mandatory.

**Comment by huoyaoyuan** [comment]
This is similar to storing syntax tree instead of source code.
Keep this in mind ultimately, the program will be stored by something close to compiled IL, together with symbol resource packs. The only additional information is difference, but it's hard to do on structured data.

**Comment by svick** [comment]
I think the biggest issue with this is tooling: How would tools like git or GitHub work in this new world? Or how would editing such a file with vim work?

A much more realistic alternative would be to have a "view" over the source code. So the file stays as it is, but when you view it in Visual Studio, you see it formatted the way you prefer. That way, all tools still keep working.

**Comment by alrz** [comment]
A similar feature in IntelliJ where you see argument names even if they are not in the original source.

**Comment by MarkusAmshove** [comment]
[MPS](https://www.jetbrains.com/mps/) is using this and call it a [projection editor](https://www.jetbrains.com/mps/concepts/index.html#projection-editor).

I do like it in a way where the code doesn't matter that much, e.g. I can see the point in specific DSLs where the user doesn't use any version control (or doesn't care).

However, as a developer I don't want to diff XML files as my source files :-)

**Comment by sharwell** [comment]
I've heard about this idea in various forms for a long time now, and conceptually I still think it's an interesting research area. However, over time I've realized that developers quickly learn to *read* source code in many different styles, and generally are not at a disadvantage as long as the body of code is consistent with itself. The painful part occurs when developers need to put effort into writing with a style that isn't their own natural/preferred style. We've used these insights to hone our strategy, with a goal of removing the difficulty of maintaining (i.e. automating) code with a project-defined style.

The ideas expressed here are really still research / far-future type ideas from the perspective of this project. Considering we are already working on an alternative strategy to address the primary motivating problems that lead to this feature request, I'm going to resolve this as **Won't Fix**. We may reopen it in the future if new empirical evidence is available that the time for change has arrived. 😄
